### ISSUE TITLE
[Bug][Compiler-v2] STLOC_UNSAFE_TO_DESTROY_ERROR

### ISSUE THREAD

**Issue by zzjas** [aptos-labs/aptos-core#13952]
# 🐛 Bug

Compiling the following code with V2:
```
module 0xCAFE::Module0 {
    struct Struct0 has drop, copy {
        x: bool,
    }

    public fun function5(var21: bool, var23: bool) {
        let var67 =  (&(var21) != &((var21 || var23)));
        Struct0 {
            x: var21
        };
    }
}
```

will raise:
```
bug: bytecode verification failed with unexpected status code `STLOC_UNSAFE_TO_DESTROY_ERROR`. This is a compiler bug, consider reporting it.
  ┌─ ~/debug_package/sources/repro.move:7:36
  │
7 │         let var67 =  (&(var21) != &((var21 || var23)));
  │                                    ^^^^^^^^^^^^^^^^^^

{
  "Error": "Move compilation failed: exiting with bytecode verification errors"
}
```

It can be compiled successfully with V1.

## Additional context

If we add `inline` for `function5`, it will compile fine:
```
module 0xCAFE::Module0 {
    struct Struct0 has drop, copy {
        x: bool,
    }

    public inline fun function5(var21: bool, var23: bool) {
        let var67 =  (&(var21) != &((var21 || var23)));
        Struct0 {
            x: var21
        };
    }
}
```

However, if we add a caller like:
```
public fun caller(a: bool, b: bool) {
       function5(a, b);
}
```
It will give the same error.

---

If the use of `var21` is removed, i.e.:

```
module 0xCAFE::Module0 {
    public fun function5(var21: bool, var23: bool) {
        let var67 =  (&(var21) != &((var21 || var23)));
    }
}
```

Compiling it will only raise a known issue:

```
error: reference safety check failed on bytecode level. This is a known issue, to be fixed later, resulting from differences between safety rules of the v1 and v2 compiler. Try to rewrite your code to workaround this problem.
  ┌─ ~/debug_package/sources/repro.move:3:36
  │
3 │         let var67 =  (&(var21) != &((var21 || var23)));
  │                                    ^^^^^^^^^^^^^^^^^^

{
  "Error": "Move compilation failed: exiting with bytecode verification errors"
}
```

**Comment by wrwg** [comment]
Can't reproduce this at head.

**Comment by wrwg** [comment]
In fact it is reproducible nut only if optimizations is off. This might be only the case since peephole opts have landed. I'll add a test case for further investigation.

**Comment by wrwg** [comment]
Fixed by #14310
